### PR TITLE
mx-puppet-discord: init at 2021-01-22

### PR DESCRIPTION
**Pull request by expipiplus1**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 119510` at c2db6eb8 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>mx-puppet-discord</li>
  </ul>
</details>
<details>
  <summary>4 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/development/node-packages/node-env.nix:409:14:
```
    |
409 |       inherit dontNpmInstall preRebuild unpackPhase buildPhase;
    |              ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

util-linux is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/node-packages/node-env.nix:401:7:
```
    |
401 |       buildInputs = [ tarWrapper python nodejs ]
    |       ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall`.

Near pkgs/development/node-packages/node-env.nix:416:7:
```
    |
416 |       installPhase = ''
    |       ^
```
</li>
    <li>warning: unused-argument

Unused argument: `runtimeShell`.
Near pkgs/servers/mx-puppet-discord/default.nix:1:47:
```
  |
1 | { stdenv, fetchFromGitHub, pkgs, lib, nodejs, runtimeShell }:
  |                                               ^
```
</li>
  </ul>
</details>

File: pkgs/servers/mx-puppet-discord/default.nix [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+  nodePackages = import ./node-composition.nix {
+    inherit pkgs nodejs;
+    inherit (stdenv.hostPlatform) system;
+  };
+
+in nodePackages.package.override {
+  inherit src;
+  nativeBuildInputs = with pkgs; [ pkgs.nodePackages.node-pre-gyp pkgconfig ];
+  buildInputs = with pkgs; [ libjpeg pixman cairo pango ];
```
Comment (SuperSandro2000) [inline pkgs/servers/mx-puppet-discord/default.nix:20]:
```suggestion
  nativeBuildInputs = [ nodePackages.node-pre-gyp pkg-config ];
  buildInputs = [ libjpeg pixman cairo pango ];
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Can we use the global nodePackages?

File: pkgs/servers/mx-puppet-discord/default.nix (hunk 1 above)
Comment (expipiplus1) [reply pkgs/servers/mx-puppet-discord/default.nix:20]: oops, what was I thinking.

**Comment by expipiplus1** [comment]
> Can we use the global nodePackages?

You mean add it to node-packages.json, and keep this derivation just to add the executable?

**Comment by expipiplus1** [comment]
> > Can we use the global nodePackages?
> 
> You mean add it to node-packages.json, and keep this derivation just to add the executable?

Had a go at doing this, but building it straight from nodePackages complains about missing `node-pre-gyp`, and overriding that adding it as a build input has mismatched versions of some dependencies, so it hasn't taken into account the lockfile. Too much effort for me to fix seeing as it's working as is.

**Comment by expipiplus1** [comment]
possible to merge this?

File: pkgs/servers/mx-puppet-discord/default.nix [hunk 2]
```diff
@@ -0,0 +1,45 @@
+{ stdenv, fetchFromGitHub, pkgs, lib, nodejs, nodePackages, pkgconfig, libjpeg
```
Comment (SuperSandro2000) [inline pkgs/servers/mx-puppet-discord/default.nix:1]:
```suggestion
{ stdenv, fetchFromGitHub, pkgs, lib, nodejs, nodePackages, pkg-config, libjpeg
```

File: pkgs/servers/mx-puppet-discord/default.nix [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  };
+
+  myNodePackages = import ./node-composition.nix {
+    inherit pkgs nodejs;
+    inherit (stdenv.hostPlatform) system;
+  };
+
+in myNodePackages.package.override {
+  inherit src;
+  nativeBuildInputs = [ nodePackages.node-pre-gyp pkgconfig ];
```
Comment (SuperSandro2000) [inline pkgs/servers/mx-puppet-discord/default.nix:20]:
```suggestion
  nativeBuildInputs = [ nodePackages.node-pre-gyp pkg-config ];
```

File: pkgs/servers/mx-puppet-discord/generate.sh [hunk 4]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env nix-shell
```
Comment (SuperSandro2000) [inline pkgs/servers/mx-puppet-discord/generate.sh:1]: Why are we not using the global nodePackages?

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/servers/mx-puppet-discord/generate.sh (hunk 4 above)
Comment (expipiplus1) [reply pkgs/servers/mx-puppet-discord/generate.sh:1]: Had a go at doing this, but building it straight from nodePackages complains about missing node-pre-gyp, and overriding that adding it as a build input has mismatched versions of some dependencies, so it hasn't taken into account the lockfile. In short, I'm unable to get this working with the global nodePackages.

File: pkgs/servers/mx-puppet-discord/node-composition.nix [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# This file has been generated by node2nix 1.9.0. Do not edit!
+
+{pkgs ? import <nixpkgs> {
```
Comment (SuperSandro2000) [inline pkgs/servers/mx-puppet-discord/node-composition.nix:3]: That probably needs to be relative.

Comment (expipiplus1) [reply pkgs/servers/mx-puppet-discord/node-composition.nix:3]: Good spot, this is the same with all the `node-composition.nix` files I checked. I've fixed it here

File: pkgs/servers/mx-puppet-discord/generate.sh (hunk 4 above)
Comment (lukegb) [reply pkgs/servers/mx-puppet-discord/generate.sh:1]: I think this is fine for leaf packages (i.e. applications). Thanks for trying with the global nodePackages set, anyway.

**Pull request merged**